binder/src/main/java/io/grpc/binder/internal/Inbound.java

```diff
@@ -0,0 +1,744 @@
+
+    ClientInbound(
+        BinderTransport transport, Attributes attributes, int callId, boolean countsForInUse) {
+      super(transport, attributes, callId);
+      this.countsForInUse = countsForInUse;
+    }
+
+    @Override
+    boolean countsForInUse() {
+      return countsForInUse;
+    }
+
+    @Override
+    @GuardedBy("this")
+    protected void handlePrefix(int flags, Parcel parcel) throws StatusException {
+      Metadata headers = MetadataHelper.readMetadata(parcel, attributes);
+      statsTraceContext.clientInboundHeaders();
+      listener.headersRead(headers);
+    }
+
+    @Override
+    @GuardedBy("this")
+    protected void handleSuffix(int flags, Parcel parcel) throws StatusException {
+      closeStatus = TransactionUtils.readStatus(flags, parcel);
+      trailers = MetadataHelper.readMetadata(parcel, attributes);
+    }
+
+    @Override
+    @GuardedBy("this")
+    protected void deliverSuffix() {
+      statsTraceContext.clientInboundTrailers(trailers);
+      statsTraceContext.streamClosed(closeStatus);
+      onDeliveryState(State.CLOSED);
+      listener.closed(closeStatus, RpcProgress.PROCESSED, trailers);
+      unregister();
+    }
+
+    @Override
+    @GuardedBy("this")
+    protected void deliverCloseAbnormal(Status status) {
+      // Assume the rpc was processed if we received anything (though we can't _really_ know).
+      RpcProgress progress = receivedData() ? RpcProgress.PROCESSED : RpcProgress.DROPPED;
+      listener.closed(status, progress, MetadataHelper.EMPTY_METADATA);
+    }
+  }
+
+  // ======================================
+  // Server-side inbound transactions.
+  static final class ServerInbound extends Inbound<ServerStreamListener> {
+
+    private final BinderTransport.BinderServerTransport serverTransport;
+
+    ServerInbound(
+        BinderTransport.BinderServerTransport transport, Attributes attributes, int callId) {
+      super(transport, attributes, callId);
+      this.serverTransport = transport;
+    }
+
+    @GuardedBy("this")
+    @Override
+    protected void handlePrefix(int flags, Parcel parcel) throws StatusException {
+      String methodName = parcel.readString();
+      Metadata headers = MetadataHelper.readMetadata(parcel, attributes);
+
+      StatsTraceContext statsTraceContext =
+          serverTransport.createStatsTraceContext(methodName, headers);
+      Outbound.ServerOutbound outbound =
+          new Outbound.ServerOutbound(serverTransport, callId, statsTraceContext);
+      ServerStream stream;
+      if ((flags & TransactionUtils.FLAG_EXPECT_SINGLE_MESSAGE) != 0) {
+        stream = new SingleMessageServerStream(this, outbound, attributes);
+      } else {
+        stream = new MultiMessageServerStream(this, outbound, attributes);
+      }
+      Status status = serverTransport.startStream(stream, methodName, headers);
+      if (status.isOk()) {
+        checkNotNull(listener); // Is it ok to assume this will happen synchronously?
```

Comment:
Yes, it is safe/important enough to make it safe. Used to serverTransportListener.streamCreated() returned the ServerStreamListener directly trying to make it "easy" for the transport. However, then we faced an issue with a stream getting killed during the streamCreated() (reentrancy!!) and it was sorta racy, so it was changed to a setter so it could be set _sooner_. https://github.com/grpc/grpc-java/commit/0e27eef1683f7e501f6bdf458f641283dab623b1

Reply:
Yeah, that's what I'd figured at the time. However the javadoc doesn't say that ServerTransportListner.streamCreated will always call stream.setListener() synchronously.